// @sanity/vision/lib/_chunks-cjs/index.js
"use strict";
var icons = require("@sanity/icons"), react = require("react"), sanity = require("sanity"), router = require("sanity/router");
const visionLocaleNamespace = "vision", visionUsEnglishLocaleBundle = sanity.defineLocaleResourceBundle({
  locale: "en-US",
  namespace: visionLocaleNamespace,
  resources: () => Promise.resolve().then(function() {
    return require("./resources.js");
  })
}), visionTool = sanity.definePlugin((options) => {
  const { name, title, icon, ...config } = options || {};
  return {
    name: "@sanity/vision",
    tools: [
      {
        name: name || "vision",
        title: title || "Vision",
        icon: icon || icons.EyeOpenIcon,
        component: react.lazy(() => Promise.resolve().then(function() {
          return require("./SanityVision.js");
        })),
        options: config,
        router: router.route.create("/*")
      }
    ],
    i18n: {
      bundles: [visionUsEnglishLocaleBundle]
    }
  };
});
exports.visionLocaleNamespace = visionLocaleNamespace;
exports.visionTool = visionTool;
