// @sanity/vision/lib/_chunks-cjs/SanityVision.js
"use strict";
var jsxRuntime = require("react/jsx-runtime"), sanity = require("sanity"), ui = require("@sanity/ui"), react = require("react"), SplitPane = require("@rexxars/react-split-pane"), icons = require("@sanity/icons"), isHotkeyEsm = require("is-hotkey-esm"), CodeMirror = require("@uiw/react-codemirror"), autocomplete = require("@codemirror/autocomplete"), commands = require("@codemirror/commands"), langJavascript = require("@codemirror/lang-javascript"), language = require("@codemirror/language"), search = require("@codemirror/search"), view = require("@codemirror/view"), highlight = require("@lezer/highlight"), color = require("@sanity/color"), styledComponents = require("styled-components"), json2Csv = require("json-2-csv"), resizeObserver = require("@juggle/resize-observer"), JSON5 = require("json5"), debounce = require("lodash/debounce.js"), index = require("./index.js"), JSONInspector = require("@rexxars/react-json-inspector"), LRU = require("quick-lru");
function _interopDefaultCompat(e) {
  return e && typeof e == "object" && "default" in e ? e : { default: e };
}
var SplitPane__default = /* @__PURE__ */ _interopDefaultCompat(SplitPane), CodeMirror__default = /* @__PURE__ */ _interopDefaultCompat(CodeMirror), JSON5__default = /* @__PURE__ */ _interopDefaultCompat(JSON5), debounce__default = /* @__PURE__ */ _interopDefaultCompat(debounce), JSONInspector__default = /* @__PURE__ */ _interopDefaultCompat(JSONInspector), LRU__default = /* @__PURE__ */ _interopDefaultCompat(LRU);
const API_VERSIONS = ["v1", "vX", "v2021-03-25", "v2021-10-21", "v2022-03-07"], [DEFAULT_API_VERSION] = API_VERSIONS.slice(-1);
function DelayedSpinner(props) {
  const [show, setShow] = react.useState(!1);
  return react.useEffect(() => {
    const timer = setTimeout(() => setShow(!0), props.delay || 500);
    return () => clearTimeout(timer);
  }, [props.delay]), show ? /* @__PURE__ */ jsxRuntime.jsx(ui.Spinner, { muted: !0, size: 4 }) : null;
}
const codemirrorExtensions = [
  [langJavascript.javascriptLanguage],
  view.lineNumbers(),
  view.highlightActiveLine(),
  view.highlightActiveLineGutter(),
  search.highlightSelectionMatches(),
  view.highlightSpecialChars(),
  language.indentOnInput(),
  language.bracketMatching(),
  autocomplete.closeBrackets(),
  commands.history(),
  view.drawSelection(),
  language.syntaxHighlighting(language.defaultHighlightStyle, { fallback: !0 }),
  view.keymap.of(
    [
      // Override the default keymap for Mod-Enter to not insert a new line, we have a custom event handler for executing queries
      { key: "Mod-Enter", run: () => !0 },
      // Add the default keymap and history keymap
      commands.defaultKeymap,
      commands.historyKeymap
    ].flat().filter(Boolean)
  )
];
function useCodemirrorTheme(theme) {
  const cmTheme = react.useMemo(() => createTheme(theme), [theme]), cmHighlight = react.useMemo(() => language.syntaxHighlighting(createHighlight(theme)), [theme]);
  return [cmTheme, cmHighlight];
}
function createTheme(theme) {
  const { color: color$1, fonts } = theme.sanity, card = color$1.card.enabled, cursor = color.hues.blue[color$1.dark ? 400 : 500].hex, selection = color.hues.gray[theme.sanity.color.dark ? 900 : 100].hex;
  return view.EditorView.theme(
    {
      "&": {
        color: card.fg,
        backgroundColor: card.bg
      },
      ".cm-content": {
        caretColor: cursor
      },
      ".cm-editor": {
        fontFamily: fonts.code.family,
        fontSize: ui.rem(fonts.code.sizes[1].fontSize),
        lineHeight: "inherit"
      },
      ".cm-cursor, .cm-dropCursor": { borderLeftColor: cursor },
      "&.cm-focused .cm-selectionBackground, .cm-selectionBackground, .cm-content ::selection": {
        backgroundColor: selection
      },
      ".cm-panels": { backgroundColor: card.bg, color: card.fg },
      ".cm-panels.cm-panels-top": { borderBottom: `2px solid ${card.border}` },
      ".cm-panels.cm-panels-bottom": { borderTop: `2px solid ${card.border}` }
    },
    { dark: color$1.dark }
  );
}
function createHighlight(theme) {
  const c = theme.sanity.color.base, s = theme.sanity.color.syntax;
  return language.HighlightStyle.define([
    { tag: highlight.tags.keyword, color: s.keyword },
    { tag: [highlight.tags.propertyName, highlight.tags.name, highlight.tags.deleted, highlight.tags.character, highlight.tags.macroName], color: s.property },
    { tag: [highlight.tags.function(highlight.tags.variableName), highlight.tags.labelName], color: s.function },
    { tag: [highlight.tags.color, highlight.tags.constant(highlight.tags.name), highlight.tags.standard(highlight.tags.name)], color: s.variable },
    { tag: [highlight.tags.definition(highlight.tags.name), highlight.tags.separator], color: s.constant },
    {
      tag: [
        highlight.tags.typeName,
        highlight.tags.className,
        highlight.tags.number,
        highlight.tags.changed,
        highlight.tags.annotation,
        highlight.tags.modifier,
        highlight.tags.self,
        highlight.tags.namespace
      ],
      color: s.number
    },
    {
      tag: [highlight.tags.operator, highlight.tags.operatorKeyword, highlight.tags.url, highlight.tags.escape, highlight.tags.regexp, highlight.tags.link, highlight.tags.special(highlight.tags.string)],
      color: s.operator
    },
    { tag: [highlight.tags.meta, highlight.tags.comment], color: s.comment },
    { tag: highlight.tags.strong, fontWeight: "bold" },
    { tag: highlight.tags.emphasis, fontStyle: "italic" },
    { tag: highlight.tags.strikethrough, textDecoration: "line-through" },
    { tag: highlight.tags.heading, fontWeight: "bold", color: s.property },
    { tag: [highlight.tags.atom, highlight.tags.bool, highlight.tags.special(highlight.tags.variableName)], color: s.boolean },
    { tag: [highlight.tags.processingInstruction, highlight.tags.string, highlight.tags.inserted], color: s.string },
    { tag: highlight.tags.invalid, color: c.fg }
  ]);
}
const EditorRoot = styledComponents.styled.div`
  width: 100%;
  box-sizing: border-box;
  height: 100%;
  overflow: hidden;
  overflow: clip;
  position: relative;
  display: flex;

  & .cm-theme {
    width: 100%;
  }

  & .cm-editor {
    height: 100%;

    font-size: 16px;
    line-height: 21px;
  }

  & .cm-line {
    padding-left: ${({ theme }) => ui.rem(theme.sanity.space[3])};
  }

  & .cm-content {
    border-right-width: ${({ theme }) => ui.rem(theme.sanity.space[4])} !important;
    padding-top: ${({ theme }) => ui.rem(theme.sanity.space[5])};
  }
`;
function VisionCodeMirror(props) {
  const sanityTheme = ui.useTheme(), theme = useCodemirrorTheme(sanityTheme);
  return /* @__PURE__ */ jsxRuntime.jsx(EditorRoot, { children: /* @__PURE__ */ jsxRuntime.jsx(CodeMirror__default.default, { basicSetup: !1, theme, extensions: codemirrorExtensions, ...props }) });
}
const PERSPECTIVES = ["raw", "previewDrafts", "published"], DEFAULT_PERSPECTIVE = PERSPECTIVES[0];
function isPerspective(p) {
  return PERSPECTIVES.includes(p);
}
function encodeQueryString(query, params = {}, options = {}) {
  const searchParams = new URLSearchParams();
  searchParams.set("query", query);
  for (const [key, value] of Object.entries(params))
    searchParams.set(`$${key}`, JSON.stringify(value));
  for (const [key, value] of Object.entries(options))
    value && searchParams.set(key, `${value}`);
  return `?${searchParams}`;
}
function getBlobUrl(content, mimeType) {
  return URL.createObjectURL(
    new Blob([content], {
      type: mimeType
    })
  );
}
function getMemoizedBlobUrlResolver(mimeType, stringEncoder) {
  return /* @__PURE__ */ (() => {
    let prevResult = "", prevContent = "";
    return (input) => {
      const content = stringEncoder(input);
      if (!(typeof content != "string" || content === ""))
        return content === prevContent || (prevContent = content, prevResult && URL.revokeObjectURL(prevResult), prevResult = getBlobUrl(content, mimeType)), prevResult;
    };
  })();
}
const getJsonBlobUrl = getMemoizedBlobUrlResolver(
  "application/json",
  (input) => JSON.stringify(input, null, 2)
), getCsvBlobUrl = getMemoizedBlobUrlResolver("text/csv", (input) => json2Csv.json2csv(Array.isArray(input) ? input : [input]).trim());
function isPlainObject(obj) {
  return !!obj && typeof obj == "object" && Object.prototype.toString.call(obj) === "[object Object]";
}
const hasLocalStorage = supportsLocalStorage(), keyPrefix = "sanityVision:";
function clearLocalStorage() {
  if (hasLocalStorage)
    for (let i = 0; i < localStorage.length; i++) {
      const key = localStorage.key(i);
      key != null && key.startsWith(keyPrefix) && localStorage.removeItem(key);
    }
}
function getLocalStorage(namespace) {
  const storageKey = `${keyPrefix}${namespace}`;
  let loadedState = null;
  return { get, set, merge };
  function get(key, defaultVal) {
    const state = ensureState();
    return typeof state[key] > "u" ? defaultVal : state[key];
  }
  function set(key, value) {
    const state = ensureState();
    return state[key] = value, localStorage.setItem(storageKey, JSON.stringify(loadedState)), value;
  }
  function merge(props) {
    const state = { ...ensureState(), ...props };
    return localStorage.setItem(storageKey, JSON.stringify(state)), state;
  }
  function ensureState() {
    return loadedState === null && (loadedState = loadState()), loadedState;
  }
  function loadState() {
    if (!hasLocalStorage)
      return {};
    try {
      const stored = JSON.parse(localStorage.getItem(storageKey) || "{}");
      return isPlainObject(stored) ? stored : {};
    } catch {
      return {};
    }
  }
}
function supportsLocalStorage() {
  const mod = "lsCheck";
  try {
    return localStorage.setItem(mod, mod), localStorage.removeItem(mod), !0;
  } catch {
    return !1;
  }
}
function parseApiQueryString(qs) {
  const params = {}, options = {};
  for (const [key, value] of qs.entries()) {
    if (key[0] === "$") {
      params[key.slice(1)] = JSON.parse(value);
      continue;
    }
    if (key === "perspective") {
      options[key] = value;
      continue;
    }
  }
  return { query: qs.get("query") || "", params, options };
}
function prefixApiVersion(version) {
  return version[0] !== "v" && version !== "other" ? `v${version}` : version;
}
const RO = typeof document > "u" ? resizeObserver.ResizeObserver : window.ResizeObserver;
function tryParseParams(val, t) {
  try {
    const parsed = val ? JSON5__default.default.parse(val) : {};
    return typeof parsed == "object" && parsed && !Array.isArray(parsed) ? parsed : {};
  } catch (err) {
    return err.message = `${t("params.error.params-invalid-json")}:

${err.message.replace(
      "JSON5:",
      ""
    )}`, err;
  }
}
function validateApiVersion(apiVersion) {
  const parseableApiVersion = apiVersion.replace(/^v/, "").trim().toUpperCase();
  return parseableApiVersion.length > 0 && (parseableApiVersion === "X" || parseableApiVersion === "1" || /^\d{4}-\d{2}-\d{2}$/.test(parseableApiVersion) && !isNaN(Date.parse(parseableApiVersion)));
}
const defaultValue = `{
  
}`;
function ParamsEditor(props) {
  const { onChange } = props, { t } = sanity.useTranslation(index.visionLocaleNamespace), { raw: value, error, parsed, valid } = eventFromValue(props.value, t), [isValid, setValid] = react.useState(valid), [init, setInit] = react.useState(!1);
  react.useEffect(() => {
    init || (onChange({ parsed, raw: value, valid: isValid, error }), setInit(!0));
  }, [error, init, isValid, onChange, parsed, value]);
  const handleChangeRaw = react.useCallback(
    (newValue) => {
      const event = eventFromValue(newValue, t);
      setValid(event.valid), onChange(event);
    },
    [onChange, t]
  ), handleChange = react.useMemo(() => debounce__default.default(handleChangeRaw, 333), [handleChangeRaw]);
  return /* @__PURE__ */ jsxRuntime.jsx(VisionCodeMirror, { value: props.value || defaultValue, onChange: handleChange });
}
function eventFromValue(value, t) {
  const parsedParams = tryParseParams(value, t), params = parsedParams instanceof Error ? {} : parsedParams, validationError = parsedParams instanceof Error ? parsedParams.message : void 0;
  return {
    parsed: params,
    raw: value,
    valid: !validationError,
    error: validationError
  };
}
const PerspectivePopoverContent = styledComponents.styled(ui.Box)`
  /* This limits the width of the popover content */
  max-width: 240px;
`, PerspectivePopoverLink = styledComponents.styled.a`
  cursor: pointer;
  margin-right: auto;
`;
function PerspectivePopover() {
  const [open, setOpen] = react.useState(!1), [buttonEl, setButtonEl] = react.useState(null), [popoverEl, setPopoverEl] = react.useState(null), handleClick = react.useCallback(() => setOpen((o) => !o), []), handleClickOutside = react.useCallback(() => setOpen(!1), []), { t } = sanity.useTranslation(index.visionLocaleNamespace);
  return ui.useClickOutside(handleClickOutside, [buttonEl, popoverEl]), /* @__PURE__ */ jsxRuntime.jsx(
    ui.Popover,
    {
      content: /* @__PURE__ */ jsxRuntime.jsx(PerspectivePopoverContent, { children: /* @__PURE__ */ jsxRuntime.jsxs(ui.Stack, { space: 4, children: [
        /* @__PURE__ */ jsxRuntime.jsxs(ui.Inline, { space: 2, children: [
          /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { weight: "medium", children: t("settings.perspectives.title") }),
          /* @__PURE__ */ jsxRuntime.jsx(ui.Badge, { tone: "primary", children: t("label.new") })
        ] }),
        /* @__PURE__ */ jsxRuntime.jsx(ui.Card, { children: /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { muted: !0, children: t("settings.perspectives.description") }) }),
        /* @__PURE__ */ jsxRuntime.jsx(ui.Card, { children: /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { children: /* @__PURE__ */ jsxRuntime.jsxs(PerspectivePopoverLink, { href: "https://sanity.io/docs/perspectives", target: "_blank", children: [
          t("settings.perspectives.action.docs-link"),
          " \u2192"
        ] }) }) })
      ] }) }),
      placement: "bottom-start",
      portal: !0,
      padding: 3,
      ref: setPopoverEl,
      open,
      children: /* @__PURE__ */ jsxRuntime.jsx(
        ui.Button,
        {
          icon: icons.HelpCircleIcon,
          mode: "bleed",
          padding: 2,
          tone: "primary",
          fontSize: 1,
          ref: setButtonEl,
          onClick: handleClick,
          selected: open
        }
      )
    }
  );
}
const ErrorCode = styledComponents.styled(ui.Code)`
  color: ${({ theme }) => theme.sanity.color.muted.critical.enabled.fg};
`;
function QueryErrorDetails({ error }) {
  const { t } = sanity.useTranslation(index.visionLocaleNamespace);
  if (!("details" in error))
    return null;
  const details = { ...error.details, ...mapToLegacyDetails(error.details) };
  return details.line ? /* @__PURE__ */ jsxRuntime.jsxs("div", { children: [
    /* @__PURE__ */ jsxRuntime.jsx(ErrorCode, { size: 1, children: `${details.line}
${dashLine(
      details.column,
      details.columnEnd
    )}` }),
    /* @__PURE__ */ jsxRuntime.jsx(ui.Box, { marginTop: 4, children: /* @__PURE__ */ jsxRuntime.jsx(ErrorCode, { size: 1, children: `${t("query.error.line")}:   ${details.lineNumber}
${t(
      "query.error.column"
    )}: ${details.column}` }) })
  ] }) : null;
}
function mapToLegacyDetails(details) {
  if (!details || typeof details.query != "string" || typeof details.start != "number")
    return {};
  const { query, start, end } = details, lineStart = query.slice(0, start).lastIndexOf(`
`) + 1, lineNumber = (query.slice(0, lineStart).match(/\n/g) || []).length, line = query.slice(lineStart, query.indexOf(`
`, lineStart)), column = start - lineStart, columnEnd = typeof end == "number" ? end - lineStart : void 0;
  return { line, lineNumber, column, columnEnd };
}
function dashLine(column, columnEnd) {
  const line = "-".repeat(column), hats = "^".repeat(columnEnd ? columnEnd - column : 1);
  return `${line}${hats}`;
}
function QueryErrorDialog(props) {
  return /* @__PURE__ */ jsxRuntime.jsxs(ui.Stack, { space: 5, marginTop: 2, children: [
    /* @__PURE__ */ jsxRuntime.jsx(ErrorCode, { size: 1, children: props.error.message }),
    /* @__PURE__ */ jsxRuntime.jsx(QueryErrorDetails, { error: props.error })
  ] });
}
const ResultViewWrapper = styledComponents.styled.div(({ theme }) => {
  const { color: color2, fonts, space } = theme.sanity;
  return styledComponents.css`
    & .json-inspector,
    & .json-inspector .json-inspector__selection {
      font-family: ${fonts.code.family};
      font-size: ${fonts.code.sizes[2].fontSize}px;
      line-height: ${fonts.code.sizes[2].lineHeight}px;
      color: var(--card-code-fg-color);
    }

    & .json-inspector .json-inspector__leaf {
      padding-left: ${ui.rem(space[4])};
    }

    & .json-inspector .json-inspector__leaf.json-inspector__leaf_root {
      padding-top: ${ui.rem(space[0])};
      padding-left: 0;
    }

    & .json-inspector > .json-inspector__leaf_root > .json-inspector__line > .json-inspector__key {
      display: none;
    }

    & .json-inspector .json-inspector__line {
      display: block;
      position: relative;
      cursor: default;
    }

    & .json-inspector .json-inspector__line::after {
      content: '';
      position: absolute;
      top: 0;
      left: -200px;
      right: -50px;
      bottom: 0;
      z-index: -1;
      pointer-events: none;
    }

    & .json-inspector .json-inspector__line:hover::after {
      background: var(--card-code-bg-color);
    }

    & .json-inspector .json-inspector__leaf_composite > .json-inspector__line {
      cursor: pointer;
    }

    & .json-inspector .json-inspector__leaf_composite > .json-inspector__line::before {
      content: '▸ ';
      margin-left: calc(0px - ${ui.rem(space[4])});
      font-size: ${fonts.code.sizes[2].fontSize}px;
      line-height: ${fonts.code.sizes[2].lineHeight}px;
    }

    &
      .json-inspector
      .json-inspector__leaf_expanded.json-inspector__leaf_composite
      > .json-inspector__line::before {
      content: '▾ ';
      font-size: ${fonts.code.sizes[2].fontSize}px;
      line-height: ${fonts.code.sizes[2].lineHeight}px;
    }

    & .json-inspector .json-inspector__radio,
    & .json-inspector .json-inspector__flatpath {
      display: none;
    }

    & .json-inspector .json-inspector__value {
      margin-left: ${ui.rem(space[4] / 2)};
    }

    &
      .json-inspector
      > .json-inspector__leaf_root
      > .json-inspector__line
      > .json-inspector__key
      + .json-inspector__value {
      margin: 0;
    }

    & .json-inspector .json-inspector__key {
      color: ${color2.syntax.property};
    }

    & .json-inspector .json-inspector__value_helper,
    & .json-inspector .json-inspector__value_null {
      color: ${color2.syntax.constant};
    }

    & .json-inspector .json-inspector__not-found {
      padding-top: ${ui.rem(space[2])};
    }

    & .json-inspector .json-inspector__value_string {
      color: ${color2.syntax.string};
      word-break: break-word;
    }

    & .json-inspector .json-inspector__value_boolean {
      color: ${color2.syntax.boolean};
    }

    & .json-inspector .json-inspector__value_number {
      color: ${color2.syntax.number};
    }

    & .json-inspector .json-inspector__show-original {
      display: inline-block;
      padding: 0 6px;
      cursor: pointer;
    }

    & .json-inspector .json-inspector__show-original:hover {
      color: inherit;
    }

    & .json-inspector .json-inspector__show-original::before {
      content: '↔';
    }

    & .json-inspector .json-inspector__show-original:hover::after {
      content: ' expand';
    }
  `;
}), lru = new LRU__default.default({ maxSize: 5e4 });
function ResultView(props) {
  const { data } = props;
  return isRecord(data) || Array.isArray(data) ? /* @__PURE__ */ jsxRuntime.jsx(ResultViewWrapper, { children: /* @__PURE__ */ jsxRuntime.jsx(
    JSONInspector__default.default,
    {
      data,
      search: !1,
      isExpanded,
      onClick: toggleExpanded
    }
  ) }) : /* @__PURE__ */ jsxRuntime.jsx(ui.Code, { language: "json", children: JSON.stringify(data) });
}
function isExpanded(keyPath, value) {
  const cached = lru.get(keyPath);
  if (typeof cached == "boolean")
    return cached;
  const segments = keyPath.split(".", 4);
  return segments.length === 4 ? !1 : Array.isArray(value) ? !0 : isRecord(value) && !segments.some((key) => isArrayKeyOverLimit(key));
}
function toggleExpanded(event) {
  const { path } = event, current = lru.get(path);
  current !== void 0 && lru.set(path, !current);
}
function isRecord(value) {
  return value !== null && typeof value == "object" && !Array.isArray(value);
}
const numeric = /^\d+$/;
function isArrayKeyOverLimit(segment, limit = 10) {
  return numeric.test(segment) && parseInt(segment, 10) > limit;
}
function preventSave(evt) {
  return evt.preventDefault();
}
function SaveCsvButton({ blobUrl }) {
  const { t } = sanity.useTranslation(index.visionLocaleNamespace), isDisabled = !blobUrl, button = /* @__PURE__ */ jsxRuntime.jsx(
    ui.Button,
    {
      as: "a",
      disabled: isDisabled,
      download: isDisabled ? void 0 : "query-result.csv",
      href: blobUrl,
      icon: icons.DocumentSheetIcon,
      mode: "ghost",
      onClick: isDisabled ? preventSave : void 0,
      text: "CSV",
      tone: "default"
    }
  );
  return isDisabled ? /* @__PURE__ */ jsxRuntime.jsx(ui.Tooltip, { content: t("result.save-result-as-csv.not-csv-encodable"), placement: "top", children: button }) : button;
}
function SaveJsonButton({ blobUrl }) {
  return /* @__PURE__ */ jsxRuntime.jsx(
    ui.Button,
    {
      as: "a",
      download: "query-result.json",
      href: blobUrl,
      icon: icons.DocumentSheetIcon,
      mode: "ghost",
      text: "JSON",
      tone: "default"
    }
  );
}
const Root = styledComponents.styled(ui.Flex)`
  .sidebarPanes .Pane {
    overflow-y: auto;
    overflow-x: hidden;
  }

  & .Resizer {
    background: var(--card-border-color);
    opacity: 1;
    z-index: 1;
    box-sizing: border-box;
    background-clip: padding-box;
    border: solid transparent;
  }

  & .Resizer:hover {
    border-color: var(--card-shadow-ambient-color);
  }

  & .Resizer.horizontal {
    height: 11px;
    margin: -5px 0;
    border-width: 5px 0;
    cursor: row-resize;
    width: 100%;
    z-index: 4;
  }

  & .Resizer.vertical {
    width: 11px;
    margin: 0 -5px;
    border-width: 0 5px;
    cursor: col-resize;
    z-index: 2; /* To prevent the resizer from being hidden behind CodeMirror scroll area */
  }

  .Resizer.disabled {
    cursor: not-allowed;
  }

  .Resizer.disabled:hover {
    border-color: transparent;
  }
`;
Root.displayName = "Root";
const Header = styledComponents.styled(ui.Card)`
  border-bottom: 1px solid var(--card-border-color);
`, StyledLabel = styledComponents.styled(ui.Label)`
  flex: 1;
`, SplitpaneContainer = styledComponents.styled(ui.Box)`
  position: relative;
`, QueryCopyLink = styledComponents.styled.a`
  cursor: pointer;
  margin-right: auto;
`, InputBackgroundContainer = styledComponents.styled(ui.Box)`
  position: absolute;
  top: 1rem;
  left: 0;
  padding: 0;
  margin: 0;
  z-index: 10;
  right: 0;

  ${StyledLabel} {
    user-select: none;
  }
`, InputBackgroundContainerLeft = styledComponents.styled(InputBackgroundContainer)`
  // This is so its aligned with the gutters of CodeMirror
  left: 33px;
`, InputContainer = styledComponents.styled(ui.Card)`
  width: 100%;
  height: 100%;
  position: relative;
  flex-direction: column;
`, ResultOuterContainer = styledComponents.styled(ui.Flex)`
  height: 100%;
`, ResultInnerContainer = styledComponents.styled(ui.Box)`
  position: relative;
`, ResultContainer = styledComponents.styled(ui.Card)`
  height: 100%;
  width: 100%;
  position: absolute;
  max-width: 100%;

  ${({ $isInvalid }) => $isInvalid && styledComponents.css`
      &:after {
        background-color: var(--card-bg-color);
        content: '';
        position: absolute;
        top: 0;
        bottom: 0;
        left: 0;
        width: 100%;
      }
    `}
`, Result = styledComponents.styled(ui.Box)`
  position: relative;
  width: 100%;
  height: 100%;
  z-index: 20;
`, ResultFooter = styledComponents.styled(ui.Flex)`
  border-top: 1px solid var(--card-border-color);
`, TimingsCard = styledComponents.styled(ui.Card)`
  position: relative;
`;
styledComponents.styled(ui.Box)`
  width: 100%;
  height: 100%;
`;
const TimingsTextContainer = styledComponents.styled(ui.Flex)`
  height: 100%;
  min-height: ${({ theme }) => ui.rem(
  theme.sanity.space[3] * 2 + theme.sanity.fonts.text.sizes[2].lineHeight - theme.sanity.fonts.text.sizes[2].ascenderHeight - theme.sanity.fonts.text.sizes[2].descenderHeight
)};
`, DownloadsCard = styledComponents.styled(ui.Card)`
  position: relative;
`, SaveResultLabel = styledComponents.styled(ui.Text)`
  transform: initial;
  &:before,
  &:after {
    content: none;
  }
  > span {
    display: flex !important;
    gap: ${({ theme }) => ui.rem(theme.sanity.space[3])};
    align-items: center;
  }
`, ControlsContainer = styledComponents.styled(ui.Box)`
  border-top: 1px solid var(--card-border-color);
`;
var __defProp$1 = Object.defineProperty, __defNormalProp$1 = (obj, key, value) => key in obj ? __defProp$1(obj, key, { enumerable: !0, configurable: !0, writable: !0, value }) : obj[key] = value, __publicField$1 = (obj, key, value) => __defNormalProp$1(obj, typeof key != "symbol" ? key + "" : key, value);
function nodeContains(node, other) {
  return !node || !other ? !1 : node === other || !!(node.compareDocumentPosition(other) & 16);
}
const sanityUrl = /\.(?:api|apicdn)\.sanity\.io\/(vX|v1|v\d{4}-\d\d-\d\d)\/.*?(?:query|listen)\/(.*?)\?(.*)/, isRunHotkey = (event) => isHotkeyEsm.isHotkey("ctrl+enter", event) || isHotkeyEsm.isHotkey("mod+enter", event);
function narrowBreakpoint() {
  return typeof window < "u" && window.innerWidth > 600;
}
function calculatePaneSizeOptions(rootHeight) {
  return {
    defaultSize: rootHeight / (narrowBreakpoint() ? 2 : 1),
    size: rootHeight > 550 ? void 0 : rootHeight * 0.4,
    allowResize: rootHeight > 550,
    minSize: Math.min(170, Math.max(170, rootHeight / 2)),
    maxSize: rootHeight > 650 ? rootHeight * 0.7 : rootHeight * 0.6
  };
}
class VisionGui extends react.PureComponent {
  constructor(props) {
    super(props), __publicField$1(this, "_visionRoot"), __publicField$1(this, "_queryEditorContainer"), __publicField$1(this, "_paramsEditorContainer"), __publicField$1(this, "_operationUrlElement"), __publicField$1(this, "_customApiVersionElement"), __publicField$1(this, "_resizeListener"), __publicField$1(this, "_querySubscription"), __publicField$1(this, "_listenSubscription"), __publicField$1(this, "_client"), __publicField$1(this, "_localStorage");
    const { client, datasets, config } = props;
    this._localStorage = getLocalStorage(client.config().projectId || "default");
    const defaultDataset = config.defaultDataset || client.config().dataset || datasets[0], defaultApiVersion = prefixApiVersion(`${config.defaultApiVersion}`), defaultPerspective = DEFAULT_PERSPECTIVE;
    let dataset = this._localStorage.get("dataset", defaultDataset), apiVersion = this._localStorage.get("apiVersion", defaultApiVersion), lastQuery = this._localStorage.get("query", ""), lastParams = this._localStorage.get("params", `{
  
}`);
    const customApiVersion = API_VERSIONS.includes(apiVersion) ? !1 : apiVersion;
    let perspective = this._localStorage.get("perspective", defaultPerspective);
    datasets.includes(dataset) || (dataset = datasets.includes(defaultDataset) ? defaultDataset : datasets[0]), API_VERSIONS.includes(apiVersion) || (apiVersion = DEFAULT_API_VERSION), PERSPECTIVES.includes(perspective) || (perspective = DEFAULT_PERSPECTIVE), typeof lastQuery != "string" && (lastQuery = ""), typeof lastParams != "string" && (lastParams = `{
  
}`), this._visionRoot = react.createRef(), this._operationUrlElement = react.createRef(), this._queryEditorContainer = react.createRef(), this._paramsEditorContainer = react.createRef(), this._customApiVersionElement = react.createRef(), this._client = props.client.withConfig({
      apiVersion: customApiVersion || apiVersion,
      dataset,
      perspective,
      allowReconfigure: !0
    });
    const bodyHeight = typeof window < "u" && typeof document < "u" ? document.body.getBoundingClientRect().height - 60 : 0, params = lastParams ? tryParseParams(lastParams, this.props.t) : void 0;
    this.state = {
      // Selected options
      dataset,
      apiVersion,
      customApiVersion,
      perspective,
      // Selected options validation state
      isValidApiVersion: customApiVersion ? validateApiVersion(customApiVersion) : !1,
      // Inputs
      query: lastQuery,
      rawParams: lastParams,
      // Parsed input
      params,
      hasValidParams: !(params instanceof Error),
      // Query/listen results
      listenMutations: [],
      // Operation state
      queryInProgress: !1,
      listenInProgress: !1,
      // UI drawing state
      paneSizeOptions: calculatePaneSizeOptions(bodyHeight)
    }, this.handleChangeDataset = this.handleChangeDataset.bind(this), this.handleChangeApiVersion = this.handleChangeApiVersion.bind(this), this.handleCustomApiVersionChange = this.handleCustomApiVersionChange.bind(this), this.handleChangePerspective = this.handleChangePerspective.bind(this), this.handleListenExecution = this.handleListenExecution.bind(this), this.handleListenerEvent = this.handleListenerEvent.bind(this), this.handleQueryExecution = this.handleQueryExecution.bind(this), this.handleQueryChange = this.handleQueryChange.bind(this), this.handleParamsChange = this.handleParamsChange.bind(this), this.handleCopyUrl = this.handleCopyUrl.bind(this), this.handlePaste = this.handlePaste.bind(this), this.handleKeyDown = this.handleKeyDown.bind(this), this.handleResize = this.handleResize.bind(this);
  }
  componentDidMount() {
    window.document.addEventListener("paste", this.handlePaste), window.document.addEventListener("keydown", this.handleKeyDown), this.handleResizeListen();
  }
  componentWillUnmount() {
    this.cancelQuery(), this.cancelListener(), this.cancelEventListener(), this.cancelResizeListener();
  }
  handleResizeListen() {
    this._visionRoot.current && (this._resizeListener = new RO(this.handleResize), this._resizeListener.observe(this._visionRoot.current));
  }
  handleResize(entries) {
    const entry = entries == null ? void 0 : entries[0];
    this.setState((prevState) => ({
      ...prevState,
      paneSizeOptions: calculatePaneSizeOptions(entry.contentRect.height)
    }));
  }
  cancelResizeListener() {
    this._resizeListener && this._resizeListener.disconnect();
  }
  handlePaste(evt) {
    if (!evt.clipboardData)
      return;
    const match = evt.clipboardData.getData("text/plain").match(sanityUrl);
    if (!match)
      return;
    const [, usedApiVersion, usedDataset, urlQuery] = match;
    let parts;
    try {
      const qs = new URLSearchParams(urlQuery);
      parts = parseApiQueryString(qs);
    } catch (err) {
      console.warn("Error while trying to parse API URL: ", err.message);
      return;
    }
    let apiVersion, customApiVersion;
    validateApiVersion(usedApiVersion) && (API_VERSIONS.includes(usedApiVersion) ? (apiVersion = usedApiVersion, customApiVersion = !1) : customApiVersion = usedApiVersion);
    const perspective = PERSPECTIVES.includes(parts.options.perspective) ? parts.options.perspective : void 0;
    evt.preventDefault(), this.setState(
      (prevState) => ({
        dataset: this.props.datasets.includes(usedDataset) ? usedDataset : prevState.dataset,
        query: parts.query,
        params: parts.params,
        rawParams: JSON.stringify(parts.params, null, 2),
        apiVersion: typeof apiVersion > "u" ? prevState.apiVersion : apiVersion,
        customApiVersion: typeof customApiVersion > "u" ? prevState.customApiVersion : customApiVersion,
        perspective: typeof perspective > "u" ? prevState.perspective : perspective
      }),
      () => {
        this._localStorage.merge({
          query: this.state.query,
          params: this.state.rawParams,
          dataset: this.state.dataset,
          apiVersion: customApiVersion || apiVersion,
          perspective: this.state.perspective
        }), this._client.config({
          dataset: this.state.dataset,
          apiVersion: customApiVersion || apiVersion,
          perspective: this.state.perspective
        }), this.handleQueryExecution(), this.props.toast.push({
          closable: !0,
          id: "vision-paste",
          status: "info",
          title: "Parsed URL to query"
        });
      }
    );
  }
  cancelQuery() {
    this._querySubscription && (this._querySubscription.unsubscribe(), this._querySubscription = void 0);
  }
  cancelListener() {
    this._listenSubscription && (this._listenSubscription.unsubscribe(), this._listenSubscription = void 0);
  }
  cancelEventListener() {
    window.removeEventListener("keydown", this.handleKeyDown);
  }
  handleChangeDataset(evt) {
    const dataset = evt.target.value;
    this._localStorage.set("dataset", dataset), this.setState({ dataset }), this._client.config({ dataset }), this.handleQueryExecution();
  }
  handleChangeApiVersion(evt) {
    const apiVersion = evt.target.value;
    if ((apiVersion == null ? void 0 : apiVersion.toLowerCase()) === "other") {
      this.setState({ customApiVersion: "v" }, () => {
        var _a;
        (_a = this._customApiVersionElement.current) == null || _a.focus();
      });
      return;
    }
    this.setState({ apiVersion, customApiVersion: !1 }, () => {
      this._localStorage.set("apiVersion", this.state.apiVersion), this._client.config({
        apiVersion: this.state.apiVersion
      }), this.handleQueryExecution();
    });
  }
  handleCustomApiVersionChange(evt) {
    const customApiVersion = evt.target.value || "", isValidApiVersion = validateApiVersion(customApiVersion);
    this.setState(
      (prevState) => ({
        apiVersion: isValidApiVersion ? customApiVersion : prevState.apiVersion,
        customApiVersion: customApiVersion || "v",
        isValidApiVersion
      }),
      () => {
        !this.state.isValidApiVersion || typeof this.state.customApiVersion != "string" || (this._localStorage.set("apiVersion", this.state.customApiVersion), this._client.config({ apiVersion: this.state.customApiVersion }));
      }
    );
  }
  handleChangePerspective(evt) {
    const perspective = evt.target.value;
    isPerspective(perspective) && this.setState({ perspective }, () => {
      this._localStorage.set("perspective", this.state.perspective), this._client.config({
        perspective: this.state.perspective
      }), this.handleQueryExecution();
    });
  }
  handleListenerEvent(evt) {
    if (evt.type !== "mutation") {
      this.props.toast.push({
        closable: !0,
        id: "vision-listen",
        status: "success",
        title: "Listening for mutations\u2026"
      });
      return;
    }
    this.setState(({ listenMutations }) => ({
      listenMutations: listenMutations.length === 50 ? [evt, ...listenMutations.slice(0, 49)] : [evt, ...listenMutations]
    }));
  }
  handleKeyDown(event) {
    const { hasValidParams } = this.state, isWithinRoot = this._visionRoot.current && nodeContains(this._visionRoot.current, event.target);
    isRunHotkey(event) && isWithinRoot && hasValidParams && (this.handleQueryExecution(), event.preventDefault(), event.stopPropagation());
  }
  ensureSelectedApiVersion() {
    const { apiVersion, customApiVersion } = this.state, wantedApiVersion = customApiVersion || apiVersion;
    this._client.config().apiVersion !== wantedApiVersion && this._client.config({ apiVersion: wantedApiVersion });
  }
  handleListenExecution() {
    const { query, params, rawParams, listenInProgress } = this.state;
    if (listenInProgress) {
      this.cancelListener(), this.setState({ listenInProgress: !1 });
      return;
    }
    this.ensureSelectedApiVersion();
    const paramsError = params instanceof Error ? params : void 0, encodeParams = params instanceof Error ? {} : params || {}, url = this._client.getDataUrl("listen", encodeQueryString(query, encodeParams, {})), shouldExecute = !paramsError && query.trim().length > 0;
    this._localStorage.set("query", query), this._localStorage.set("params", rawParams), this.cancelQuery(), this.setState({
      url,
      listenMutations: [],
      queryInProgress: !1,
      queryResult: void 0,
      listenInProgress: shouldExecute,
      error: paramsError,
      queryTime: void 0,
      e2eTime: void 0
    }), shouldExecute && (this._listenSubscription = this._client.listen(query, params, { events: ["mutation", "welcome"] }).subscribe({
      next: this.handleListenerEvent,
      error: (error) => this.setState({
        error,
        query,
        listenInProgress: !1
      })
    }));
  }
  handleQueryExecution() {
    const { query, params, rawParams, queryInProgress } = this.state;
    if (queryInProgress)
      return this.cancelQuery(), this.cancelListener(), this.setState({ queryInProgress: !1 }), !0;
    const paramsError = params instanceof Error && params;
    if (this._localStorage.set("query", query), this._localStorage.set("params", rawParams), this.cancelListener(), this.setState({
      queryInProgress: !paramsError && !!query,
      listenInProgress: !1,
      listenMutations: [],
      error: paramsError || void 0,
      queryResult: void 0,
      queryTime: void 0,
      e2eTime: void 0
    }), !query || paramsError)
      return !0;
    this.ensureSelectedApiVersion();
    const urlQueryOpts = {};
    this.state.perspective !== "raw" && (urlQueryOpts.perspective = this.state.perspective);
    const url = this._client.getUrl(
      this._client.getDataUrl("query", encodeQueryString(query, params, urlQueryOpts))
    );
    this.setState({ url });
    const queryStart = Date.now();
    return this._querySubscription = this._client.observable.fetch(query, params, { filterResponse: !1, tag: "vision" }).subscribe({
      next: (res) => this.setState({
        queryTime: res.ms,
        e2eTime: Date.now() - queryStart,
        queryResult: res.result,
        queryInProgress: !1,
        error: void 0
      }),
      error: (error) => this.setState({
        error,
        query,
        queryInProgress: !1
      })
    }), !0;
  }
  handleQueryChange(query) {
    this.setState({ query });
  }
  handleParamsChange({ raw, parsed, valid, error }) {
    this.setState(
      {
        rawParams: raw,
        params: parsed,
        hasValidParams: valid,
        paramsError: error
      },
      () => this._localStorage.set("params", raw)
    );
  }
  handleCopyUrl() {
    const el = this._operationUrlElement.current;
    if (el)
      try {
        el.select(), document.execCommand("copy"), this.props.toast.push({
          closable: !0,
          title: "Copied to clipboard",
          status: "info",
          id: "vision-copy"
        });
      } catch {
        console.error("Unable to copy to clipboard :(");
      }
  }
  render() {
    const { datasets, t } = this.props, {
      apiVersion,
      customApiVersion,
      dataset,
      e2eTime,
      error,
      hasValidParams,
      isValidApiVersion,
      listenInProgress,
      listenMutations,
      paneSizeOptions,
      paramsError,
      perspective,
      query,
      queryInProgress,
      queryResult,
      queryTime,
      rawParams,
      url
    } = this.state, hasResult = !error && !queryInProgress && typeof queryResult < "u", jsonUrl = hasResult ? getJsonBlobUrl(queryResult) : "", csvUrl = hasResult ? getCsvBlobUrl(queryResult) : "";
    return /* @__PURE__ */ jsxRuntime.jsxs(
      Root,
      {
        direction: "column",
        height: "fill",
        ref: this._visionRoot,
        sizing: "border",
        overflow: "hidden",
        children: [
          /* @__PURE__ */ jsxRuntime.jsx(Header, { paddingX: 3, paddingY: 2, children: /* @__PURE__ */ jsxRuntime.jsxs(ui.Grid, { columns: [1, 4, 8, 12], children: [
            /* @__PURE__ */ jsxRuntime.jsx(ui.Box, { padding: 1, column: 2, children: /* @__PURE__ */ jsxRuntime.jsxs(ui.Stack, { children: [
              /* @__PURE__ */ jsxRuntime.jsx(ui.Card, { paddingTop: 2, paddingBottom: 3, children: /* @__PURE__ */ jsxRuntime.jsx(StyledLabel, { children: t("settings.dataset-label") }) }),
              /* @__PURE__ */ jsxRuntime.jsx(ui.Select, { value: dataset, onChange: this.handleChangeDataset, children: datasets.map((ds) => /* @__PURE__ */ jsxRuntime.jsx("option", { children: ds }, ds)) })
            ] }) }),
            /* @__PURE__ */ jsxRuntime.jsx(ui.Box, { padding: 1, column: 2, children: /* @__PURE__ */ jsxRuntime.jsxs(ui.Stack, { children: [
              /* @__PURE__ */ jsxRuntime.jsx(ui.Card, { paddingTop: 2, paddingBottom: 3, children: /* @__PURE__ */ jsxRuntime.jsx(StyledLabel, { children: t("settings.api-version-label") }) }),
              /* @__PURE__ */ jsxRuntime.jsxs(
                ui.Select,
                {
                  value: customApiVersion === !1 ? apiVersion : "other",
                  onChange: this.handleChangeApiVersion,
                  children: [
                    API_VERSIONS.map((version) => /* @__PURE__ */ jsxRuntime.jsx("option", { children: version }, version)),
                    /* @__PURE__ */ jsxRuntime.jsx("option", { value: "other", children: t("settings.other-api-version-label") }, "other")
                  ]
                }
              )
            ] }) }),
            customApiVersion !== !1 && /* @__PURE__ */ jsxRuntime.jsx(ui.Box, { padding: 1, column: 2, children: /* @__PURE__ */ jsxRuntime.jsxs(ui.Stack, { children: [
              /* @__PURE__ */ jsxRuntime.jsx(ui.Card, { paddingTop: 2, paddingBottom: 3, children: /* @__PURE__ */ jsxRuntime.jsx(StyledLabel, { textOverflow: "ellipsis", children: t("settings.custom-api-version-label") }) }),
              /* @__PURE__ */ jsxRuntime.jsx(
                ui.TextInput,
                {
                  ref: this._customApiVersionElement,
                  value: customApiVersion,
                  onChange: this.handleCustomApiVersionChange,
                  customValidity: isValidApiVersion ? void 0 : t("settings.error.invalid-api-version"),
                  maxLength: 11
                }
              )
            ] }) }),
            /* @__PURE__ */ jsxRuntime.jsx(ui.Box, { padding: 1, column: 2, children: /* @__PURE__ */ jsxRuntime.jsxs(ui.Stack, { children: [
              /* @__PURE__ */ jsxRuntime.jsx(ui.Card, { paddingBottom: 1, children: /* @__PURE__ */ jsxRuntime.jsxs(ui.Inline, { space: 1, children: [
                /* @__PURE__ */ jsxRuntime.jsx(ui.Box, { children: /* @__PURE__ */ jsxRuntime.jsx(StyledLabel, { children: t("settings.perspective-label") }) }),
                /* @__PURE__ */ jsxRuntime.jsx(ui.Box, { children: /* @__PURE__ */ jsxRuntime.jsx(PerspectivePopover, {}) })
              ] }) }),
              /* @__PURE__ */ jsxRuntime.jsx(ui.Select, { value: perspective, onChange: this.handleChangePerspective, children: PERSPECTIVES.map((p) => /* @__PURE__ */ jsxRuntime.jsx("option", { children: p }, p)) })
            ] }) }),
            typeof url == "string" ? /* @__PURE__ */ jsxRuntime.jsx(ui.Box, { padding: 1, flex: 1, column: customApiVersion === !1 ? 6 : 4, children: /* @__PURE__ */ jsxRuntime.jsxs(ui.Stack, { children: [
              /* @__PURE__ */ jsxRuntime.jsx(ui.Card, { paddingTop: 2, paddingBottom: 3, children: /* @__PURE__ */ jsxRuntime.jsxs(StyledLabel, { children: [
                t("query.url"),
                "\xA0",
                /* @__PURE__ */ jsxRuntime.jsxs(QueryCopyLink, { onClick: this.handleCopyUrl, children: [
                  "[",
                  t("action.copy-url-to-clipboard"),
                  "]"
                ] })
              ] }) }),
              /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { flex: 1, gap: 1, children: [
                /* @__PURE__ */ jsxRuntime.jsx(ui.Box, { flex: 1, children: /* @__PURE__ */ jsxRuntime.jsx(ui.TextInput, { readOnly: !0, type: "url", ref: this._operationUrlElement, value: url }) }),
                /* @__PURE__ */ jsxRuntime.jsx(ui.Tooltip, { content: t("action.copy-url-to-clipboard"), children: /* @__PURE__ */ jsxRuntime.jsx(
                  ui.Button,
                  {
                    "aria-label": t("action.copy-url-to-clipboard"),
                    type: "button",
                    mode: "ghost",
                    icon: icons.CopyIcon,
                    onClick: this.handleCopyUrl
                  }
                ) })
              ] })
            ] }) }) : /* @__PURE__ */ jsxRuntime.jsx(ui.Box, { flex: 1 })
          ] }) }),
          /* @__PURE__ */ jsxRuntime.jsx(SplitpaneContainer, { flex: "auto", children: /* @__PURE__ */ jsxRuntime.jsxs(
            SplitPane__default.default,
            {
              split: narrowBreakpoint() ? "vertical" : "horizontal",
              minSize: 280,
              defaultSize: 400,
              maxSize: -400,
              children: [
                /* @__PURE__ */ jsxRuntime.jsx(ui.Box, { height: "stretch", flex: 1, children: /* @__PURE__ */ jsxRuntime.jsxs(
                  SplitPane__default.default,
                  {
                    className: "sidebarPanes",
                    split: "horizontal",
                    defaultSize: narrowBreakpoint() ? paneSizeOptions.defaultSize : paneSizeOptions.minSize,
                    size: paneSizeOptions.size,
                    allowResize: paneSizeOptions.allowResize,
                    minSize: narrowBreakpoint() ? paneSizeOptions.minSize : 100,
                    maxSize: paneSizeOptions.maxSize,
                    primary: "first",
                    children: [
                      /* @__PURE__ */ jsxRuntime.jsx(InputContainer, { display: "flex", ref: this._queryEditorContainer, children: /* @__PURE__ */ jsxRuntime.jsxs(ui.Box, { flex: 1, children: [
                        /* @__PURE__ */ jsxRuntime.jsx(InputBackgroundContainerLeft, { children: /* @__PURE__ */ jsxRuntime.jsx(ui.Flex, { children: /* @__PURE__ */ jsxRuntime.jsx(StyledLabel, { muted: !0, children: t("query.label") }) }) }),
                        /* @__PURE__ */ jsxRuntime.jsx(VisionCodeMirror, { value: query, onChange: this.handleQueryChange })
                      ] }) }),
                      /* @__PURE__ */ jsxRuntime.jsxs(InputContainer, { display: "flex", ref: this._paramsEditorContainer, children: [
                        /* @__PURE__ */ jsxRuntime.jsxs(ui.Card, { flex: 1, tone: hasValidParams ? "default" : "critical", children: [
                          /* @__PURE__ */ jsxRuntime.jsx(InputBackgroundContainerLeft, { children: /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { children: [
                            /* @__PURE__ */ jsxRuntime.jsx(StyledLabel, { muted: !0, children: t("params.label") }),
                            paramsError && /* @__PURE__ */ jsxRuntime.jsx(ui.Tooltip, { placement: "top-end", portal: !0, content: paramsError, children: /* @__PURE__ */ jsxRuntime.jsx(ui.Box, { padding: 1, marginX: 2, children: /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { children: /* @__PURE__ */ jsxRuntime.jsx(icons.ErrorOutlineIcon, {}) }) }) })
                          ] }) }),
                          /* @__PURE__ */ jsxRuntime.jsx(ParamsEditor, { value: rawParams, onChange: this.handleParamsChange })
                        ] }),
                        /* @__PURE__ */ jsxRuntime.jsx(ControlsContainer, { children: /* @__PURE__ */ jsxRuntime.jsx(ui.Card, { padding: 3, paddingX: 3, children: /* @__PURE__ */ jsxRuntime.jsx(
                          ui.Tooltip,
                          {
                            content: /* @__PURE__ */ jsxRuntime.jsx(ui.Card, { radius: 4, children: /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: 1, muted: !0, children: t("params.error.params-invalid-json") }) }),
                            placement: "top",
                            disabled: hasValidParams,
                            portal: !0,
                            children: /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { justify: "space-evenly", children: [
                              /* @__PURE__ */ jsxRuntime.jsx(ui.Box, { flex: 1, children: /* @__PURE__ */ jsxRuntime.jsx(
                                ui.Tooltip,
                                {
                                  content: /* @__PURE__ */ jsxRuntime.jsx(ui.Card, { radius: 4, children: /* @__PURE__ */ jsxRuntime.jsx(ui.Hotkeys, { keys: ["Ctrl", "Enter"] }) }),
                                  placement: "top",
                                  portal: !0,
                                  children: /* @__PURE__ */ jsxRuntime.jsx(
                                    ui.Button,
                                    {
                                      width: "fill",
                                      onClick: this.handleQueryExecution,
                                      type: "button",
                                      icon: queryInProgress ? icons.StopIcon : icons.PlayIcon,
                                      disabled: listenInProgress || !hasValidParams,
                                      tone: queryInProgress ? "positive" : "primary",
                                      text: t(queryInProgress ? "action.query-cancel" : "action.query-execute")
                                    }
                                  )
                                }
                              ) }),
                              /* @__PURE__ */ jsxRuntime.jsx(ui.Box, { flex: 1, marginLeft: 3, children: /* @__PURE__ */ jsxRuntime.jsx(
                                ui.Button,
                                {
                                  width: "fill",
                                  onClick: this.handleListenExecution,
                                  type: "button",
                                  icon: listenInProgress ? icons.StopIcon : icons.PlayIcon,
                                  text: t(listenInProgress ? "action.listen-cancel" : "action.listen-execute"),
                                  mode: "ghost",
                                  disabled: !hasValidParams,
                                  tone: listenInProgress ? "positive" : "default"
                                }
                              ) })
                            ] })
                          }
                        ) }) })
                      ] })
                    ]
                  }
                ) }),
                /* @__PURE__ */ jsxRuntime.jsxs(ResultOuterContainer, { direction: "column", children: [
                  /* @__PURE__ */ jsxRuntime.jsx(ResultInnerContainer, { flex: 1, children: /* @__PURE__ */ jsxRuntime.jsx(
                    ResultContainer,
                    {
                      flex: 1,
                      overflow: "hidden",
                      tone: error ? "critical" : "default",
                      $isInvalid: !!error,
                      children: /* @__PURE__ */ jsxRuntime.jsxs(Result, { overflow: "auto", children: [
                        /* @__PURE__ */ jsxRuntime.jsx(InputBackgroundContainer, { children: /* @__PURE__ */ jsxRuntime.jsx(ui.Box, { marginLeft: 3, children: /* @__PURE__ */ jsxRuntime.jsx(StyledLabel, { muted: !0, children: t("result.label") }) }) }),
                        /* @__PURE__ */ jsxRuntime.jsxs(ui.Box, { padding: 3, paddingTop: 5, children: [
                          (queryInProgress || listenInProgress && listenMutations.length === 0) && /* @__PURE__ */ jsxRuntime.jsx(ui.Box, { marginTop: 3, children: /* @__PURE__ */ jsxRuntime.jsx(DelayedSpinner, {}) }),
                          error && /* @__PURE__ */ jsxRuntime.jsx(QueryErrorDialog, { error }),
                          hasResult && /* @__PURE__ */ jsxRuntime.jsx(ResultView, { data: queryResult }),
                          listenInProgress && listenMutations.length > 0 && /* @__PURE__ */ jsxRuntime.jsx(ResultView, { data: listenMutations })
                        ] })
                      ] })
                    }
                  ) }),
                  /* @__PURE__ */ jsxRuntime.jsxs(ResultFooter, { justify: "space-between", direction: ["column", "column", "row"], children: [
                    /* @__PURE__ */ jsxRuntime.jsx(TimingsCard, { paddingX: 4, paddingY: 3, sizing: "border", children: /* @__PURE__ */ jsxRuntime.jsxs(TimingsTextContainer, { align: "center", children: [
                      /* @__PURE__ */ jsxRuntime.jsx(ui.Box, { children: /* @__PURE__ */ jsxRuntime.jsxs(ui.Text, { muted: !0, children: [
                        t("result.execution-time-label"),
                        ":",
                        " ",
                        typeof queryTime == "number" ? `${queryTime}ms` : t("result.timing-not-applicable")
                      ] }) }),
                      /* @__PURE__ */ jsxRuntime.jsx(ui.Box, { marginLeft: 4, children: /* @__PURE__ */ jsxRuntime.jsxs(ui.Text, { muted: !0, children: [
                        t("result.end-to-end-time-label"),
                        ":",
                        " ",
                        typeof e2eTime == "number" ? `${e2eTime}ms` : t("result.timing-not-applicable")
                      ] }) })
                    ] }) }),
                    hasResult && /* @__PURE__ */ jsxRuntime.jsx(DownloadsCard, { paddingX: 4, paddingY: 3, sizing: "border", children: /* @__PURE__ */ jsxRuntime.jsx(SaveResultLabel, { muted: !0, children: /* @__PURE__ */ jsxRuntime.jsx(
                      sanity.Translate,
                      {
                        components: {
                          SaveResultButtons: () => /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
                            /* @__PURE__ */ jsxRuntime.jsx(SaveJsonButton, { blobUrl: jsonUrl }),
                            /* @__PURE__ */ jsxRuntime.jsx(SaveCsvButton, { blobUrl: csvUrl })
                          ] })
                        },
                        i18nKey: "result.save-result-as-format",
                        t
                      }
                    ) }) })
                  ] })
                ] })
              ]
            }
          ) })
        ]
      }
    );
  }
}
function useDatasets(client) {
  const projectId = client.config().projectId, [datasets, setDatasets] = react.useState();
  return react.useEffect(() => {
    const datasets$ = client.observable.datasets.list().subscribe({
      next: (result) => setDatasets(result.map((ds) => ds.name)),
      error: (err) => setDatasets(err)
    });
    return () => datasets$.unsubscribe();
  }, [client, projectId]), datasets || void 0;
}
function VisionContainer(props) {
  const toast = ui.useToast(), loadedDatasets = useDatasets(props.client), { t } = sanity.useTranslation(index.visionLocaleNamespace);
  if (!loadedDatasets)
    return /* @__PURE__ */ jsxRuntime.jsx(DelayedSpinner, {});
  const datasets = loadedDatasets instanceof Error ? (
    // On error, use the clients configured dataset
    [props.client.config().dataset || "production"]
  ) : (
    // Otherwise use the loaded list, obviously
    loadedDatasets
  );
  return /* @__PURE__ */ jsxRuntime.jsx(VisionGui, { ...props, datasets, toast, t });
}
var __defProp = Object.defineProperty, __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: !0, configurable: !0, writable: !0, value }) : obj[key] = value, __publicField = (obj, key, value) => __defNormalProp(obj, typeof key != "symbol" ? key + "" : key, value);
class VisionErrorBoundary extends react.Component {
  constructor(props) {
    super(props), __publicField(this, "handleRetryRender", () => this.setState((prev) => ({ error: null, numRetries: prev.numRetries + 1 }))), __publicField(this, "handleRetryWithCacheClear", () => {
      clearLocalStorage(), this.handleRetryRender();
    }), this.state = { error: null, numRetries: 0 };
  }
  static getDerivedStateFromError(error) {
    return { error: error instanceof Error ? error.message : `${error}` };
  }
  render() {
    if (!this.state.error)
      return this.props.children;
    const message = this.state.error, withCacheClear = this.state.numRetries > 0;
    return /* @__PURE__ */ jsxRuntime.jsx(
      ui.Card,
      {
        height: "fill",
        overflow: "auto",
        paddingY: [4, 5, 6, 7],
        paddingX: 4,
        sizing: "border",
        tone: "critical",
        children: /* @__PURE__ */ jsxRuntime.jsx(ui.Container, { width: 3, children: /* @__PURE__ */ jsxRuntime.jsxs(ui.Stack, { space: 4, children: [
          /* @__PURE__ */ jsxRuntime.jsx("div", { children: /* @__PURE__ */ jsxRuntime.jsx(
            ui.Button,
            {
              onClick: withCacheClear ? this.handleRetryWithCacheClear : this.handleRetryRender,
              text: withCacheClear ? "Clear cache and retry" : "Retry",
              tone: "default"
            }
          ) }),
          /* @__PURE__ */ jsxRuntime.jsx(ui.Heading, { children: "An error occured" }),
          /* @__PURE__ */ jsxRuntime.jsx(ui.Card, { border: !0, radius: 2, overflow: "auto", padding: 4, tone: "inherit", children: /* @__PURE__ */ jsxRuntime.jsx(ui.Stack, { space: 4, children: message && /* @__PURE__ */ jsxRuntime.jsx(ui.Code, { size: 1, children: /* @__PURE__ */ jsxRuntime.jsxs("strong", { children: [
            "Error: ",
            message
          ] }) }) }) })
        ] }) })
      }
    );
  }
}
function SanityVision(props) {
  const client = sanity.useClient({ apiVersion: "1" }), config = {
    defaultApiVersion: DEFAULT_API_VERSION,
    ...props.tool.options
  };
  return /* @__PURE__ */ jsxRuntime.jsx(VisionErrorBoundary, { children: /* @__PURE__ */ jsxRuntime.jsx(VisionContainer, { client, config }) });
}
exports.default = SanityVision;
